File: apps/api/src/model-provider.ts
import {
  type Model,
  OpenAIResponsesModel,
  setTracingDisabled,
} from "@openai/agents";
import OpenAI from "openai";

import { loadSettings, type Settings } from "./settings.ts";

export class MissingCredentialsError extends Error {}

/** Agent 层只需要“用哪个 Model”和“凭据是否齐全”，因此依赖这个接口而非具体实现。 */
export type ModelProvider = {
  readonly agentModel: Model;
  readonly configured: boolean;
  readonly modelId: string;
};

/** Task Framer 和主 Agent Loop 共用同一个 DeepSeek Responses 客户端。 */
export class ResponsesModelProvider implements ModelProvider {
  readonly modelId: string;
  readonly configured: boolean;
  readonly agentModel: Model;

  constructor(settings: Settings = loadSettings()) {
    this.modelId = settings.model;
    this.configured = Boolean(settings.apiKey);
    const client = new OpenAI({
      apiKey: settings.apiKey || "not-configured",
      baseURL: settings.baseUrl,
    });
    this.agentModel = new OpenAIResponsesModel(client, this.modelId);
    // DeepSeek 不接收 OpenAI trace，因此关闭 SDK 默认 tracing。
    setTracingDisabled(true);
  }
}

File: backend/app/model_provider.py
from __future__ import annotations

from agents import OpenAIResponsesModel, set_tracing_disabled
from openai import AsyncOpenAI

from app.settings import Settings, load_settings


class MissingCredentialsError(Exception):
    pass


class ResponsesModelProvider:
    """Task Framer 和主 Agent Loop 共用同一个 DeepSeek Responses 客户端。"""

    def __init__(self, settings: Settings | None = None) -> None:
        resolved_settings = settings or load_settings()
        self.model_id = resolved_settings.model
        self.configured = bool(resolved_settings.api_key)
        self.client = AsyncOpenAI(
            api_key=resolved_settings.api_key or "not-configured",
            base_url=resolved_settings.base_url,
        )
        self.agent_model = OpenAIResponsesModel(
            model=self.model_id,
            openai_client=self.client,
        )
        # DeepSeek 不接收 OpenAI trace，因此关闭 SDK 默认 tracing。
        set_tracing_disabled(True)

    async def close(self) -> None:
        await self.client.close()
